src/System.Memory/src/System/Buffers/StandardFormat.cs

```diff
@@ -0,0 +1,123 @@
+        /// The maximum valid precision value.
+        /// </summary>
+        public const byte NoPrecision = byte.MaxValue;
+
+        /// <summary>
+        /// Precision values for format that don't use a precision, or for when the precision is to be unspecified.
+        /// </summary>
+        public const byte MaxPrecision = 99;
+
+        private readonly byte _format;
+        private readonly byte _precision;
+
+        /// <summary>
+        /// The character component of the format.
+        /// </summary>
+        public char Symbol => (char)_format;
+
+        /// <summary>
+        /// The precision component of the format. Ranges from 0..9 or the special value NoPrecision.
+        /// </summary>
+        public byte Precision => _precision;
+
+        /// <summary>
+        /// true if Precision is a value other than NoPrecision
+        /// </summary>
+        public bool HasPrecision => _precision != NoPrecision;
+
+        /// <summary>
+        /// true if the StandardFormat == default(StandardFormat)
+        /// </summary>
+        public bool IsDefault => _format == 0 && _precision == 0;
+
+        /// <summary>
+        /// Create a StandardFormat.
+        /// </summary>
+        /// <param name="symbol">A type-specific formatting characeter such as 'G', 'D' or 'X'</param>
+        /// <param name="precision">An optional precision ranging from 0..9 or the special value NoPrecision (the default)</param>
+        public StandardFormat(char symbol, byte precision = NoPrecision)
+        {
+            if (precision != NoPrecision && precision > MaxPrecision)
+                throw new ArgumentOutOfRangeException(nameof(precision));
+            if (symbol != (byte)symbol)
+                throw new ArgumentOutOfRangeException(nameof(symbol));
+
+            _format = (byte)symbol;
+            _precision = precision;
+        }
+
+        /// <summary>
+        /// Converts a character to a StandardFormat using the NoPrecision precision.
+        /// </summary>
+        public static implicit operator StandardFormat(char symbol) => new StandardFormat(symbol);
+
+        /// <summary>
+        /// Converts a classic .NET format string into a StandardFormat
+        /// </summary>
+        public static StandardFormat Parse(ReadOnlySpan<char> format)
+        {
+            return Parse(new string(format.ToArray()));
+        }
+
+        /// <summary>
+        /// Converts a classic .NET format string into a StandardFormat
+        /// </summary>
+        public static StandardFormat Parse(string format)
+        {
+            if (string.IsNullOrEmpty(format))
+                return default;
+
+            char specifier = format[0];
+            byte precision = NoPrecision;
+
+            if (format.Length > 1)
+            {
+                if (!byte.TryParse(format.Substring(1), out precision))
+                    throw new FormatException("format");
+
+                if (precision > MaxPrecision)
```

Comment:
`NoPrecision` is already an illegal value that's greater than `MaxPrecision`. So the one check covers both.

Reply:
Hmm. So, right now this is the behavior:
Parse("G255") => Format Exception
Parse("G") => valid StandardFormat where StandardFormat.Precision == 255

Is the suggested pattern to always check StandardFormat.HasPrecision before accessing the Precision property? And the correct way to create StandardFormat without precision is to just not specify one?

Should NoPrecision be private?

Reply:
The `Parse` method is there to convert classic BCL format strings - since "G255" != "G" there, it doesn't make sense to accept it.

>Is the suggested pattern to always check StandardFormat.HasPrecision before accessing the Precision property? 

It's an extra check at runtime and this is a perf-sensitive area, so while it reads nicely and I wouldn't frown on it for non-perf-critical paths, it probably should be a soft guidance at best.

For awareness, I think it makes sense to make `NoPrecision` public especially given that its runtime value is not the same as `default(StandardFormat)` It hammers in the fact that `NoPrecision` and `0` mean different things here.